src/main/java/com/google/step/snippet/servlets/FeedbackServlet.java

```diff
@@ -0,0 +1,174 @@
+
+    if (userService.isUserLoggedIn()) {
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      String url = request.getParameter(URL);
+      Entity userEntity = getUserEntity(datastore, userService);
+
+      // If a user has previously voted on this card, retrieve their past voting status
+      if (url != null) {
+        ArrayList<String> userUpCards = (ArrayList<String>) userEntity.getProperty(UPVOTED);
+        ArrayList<String> userDownCards = (ArrayList<String>) userEntity.getProperty(DOWNVOTED);
+        if (userUpCards != null && userUpCards.contains(url)) {
+          json.put(STATUS, UPVOTED);
+        } else if (userDownCards != null && userDownCards.contains(url)) {
+          json.put(STATUS, DOWNVOTED);
+        }
+      }
+    }
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    // Invalid POST request if URL is invalid or user is not logged
+    String url = request.getParameter(URL);
+    UserService userService = UserServiceFactory.getUserService();
+    if (url == null || !userService.isUserLoggedIn()) {
+      JSONObject json = new JSONObject();
+      json.put(STATUS, NONE);
+      response.setContentType("application/json");
+      response.getWriter().println(json);
+      return;
+    }
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity userEntity = getUserEntity(datastore, userService);
+
+    // Get lists of user's upvoted and downvoted cards
+    ArrayList<String> upCards;
+    if (userEntity.getProperty(UPVOTED) == null) {
+      upCards = new ArrayList<String>();
+    } else {
+      upCards = (ArrayList<String>) userEntity.getProperty(UPVOTED);
+    }
+    ArrayList<String> downCards;
+    if (userEntity.getProperty(DOWNVOTED) == null) {
+      downCards = new ArrayList<String>();
+    } else {
+      downCards = (ArrayList<String>) userEntity.getProperty(DOWNVOTED);
+    }
+
+    // Add or remove card from user's list of upvoted and downvoted cards
+    if (request.getParameter(UP) != null) {
+      if (!upCards.contains(url)) {
+        upCards.add(url);
+        if (downCards.contains(url)) {
+          downCards.remove(url);
+        }
+      } else {
+        upCards.remove(url);
+      }
+    } else if (request.getParameter(DOWN) != null) {
+      if (!downCards.contains(url)) {
+        downCards.add(url);
+        if (upCards.contains(url)) {
+          upCards.remove(url);
+        }
+      } else {
+        downCards.remove(url);
+      }
+    }
+    userEntity.setProperty(UPVOTED, upCards);
+    userEntity.setProperty(DOWNVOTED, downCards);
+    datastore.put(userEntity);
+
+    // Retrieve toggle status of upvote and downvote buttons for user
+    String status = NONE;
+    if (userEntity.getProperty(UPVOTED) != null
```

Comment:
How about you just look at upCards? It's the same thing, and you don't have to do null checks. (Incidentally, this will never be null since you just set it to be a non-null thing on line 124. Something something proving safety of a null dereference.)